tool/ci/pkg/environment/environment.go

```diff
@@ -0,0 +1,289 @@
+		for _, reviewer := range requiredReviewers {
+			_, err := userExists(reviewer, client)
+			if err != nil {
+				return nil, trace.Wrap(err)
+			}
+		}
+		if author == "" {
+			hasDefaultReviewers = true
+			continue
+		}
+		_, err := userExists(author, client)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+	}
+	if !hasDefaultReviewers {
+		return nil, trace.BadParameter("default reviewers are not set. set default reviewers with an empty string as a key")
+	}
+	return m, nil
+
+}
+
+// userExists checks if a user exists
+func userExists(userLogin string, client *github.Client) (*github.User, error) {
+	var targetUser *github.User
+	users, resp, err := client.Search.Users(context.TODO(), userLogin, &github.SearchOptions{})
+	if err != nil || resp.StatusCode != http.StatusOK {
+		return nil, trace.Wrap(err)
+	}
+	for _, user := range users.Users {
+		if *user.Login == userLogin {
+			targetUser = user
+		}
+	}
+	if targetUser == nil {
+		return nil, trace.NotFound("user %s does not exist", userLogin)
+	}
+	return targetUser, nil
+}
+
+// GetReviewersForAuthor gets the required reviewers for the current user.
+func (e *Environment) GetReviewersForAuthor(user string) []string {
+	value, ok := e.reviewers[user]
+	// author is external or does not have set reviewers
+	if !ok {
+		return e.defaultReviewers
+	}
+	return value
+}
+
+// IsInternal determines if an author is an internal contributor.
+func (e *Environment) IsInternal(author string) bool {
+	_, ok := e.reviewers[author]
+	return ok
+}
+
+// GetToken gets token
+func (e *Environment) GetToken() string {
+	return e.token
+}
+
+// GetPullRequest gets the pull request metadata in the current context.
+func GetPullRequest(path string) (*PullRequestMetadata, error) {
+	var actionType action
+	file, err := os.Open(path)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	defer file.Close()
+	body, err := ioutil.ReadAll(file)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	err = json.Unmarshal(body, &actionType)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return getPullRequest(body, actionType.Action)
+}
+
+func getPullRequest(body []byte, action string) (*PullRequestMetadata, error) {
+	var pr PullRequestMetadata
+
+	switch action {
+	case ci.Synchronize:
+		var push PushEvent
+		err := json.Unmarshal(body, &push)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		pr, err = push.toPullRequestMetadata()
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+	case ci.Assigned, ci.Opened, ci.Reopened, ci.Ready:
+		var pull PullRequestEvent
+		err := json.Unmarshal(body, &pull)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		pr, err = pull.toPullRequestMetadata()
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+	default:
+		var rev ReviewEvent
+		err := json.Unmarshal(body, &rev)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		pr, err = rev.toPullRequestMetadata()
+		if err != nil {
+			return nil, err
+		}
+	}
```

Comment:
Seems to be a lot of overlap in these structs, any reason you chose to break them out like this?

Reply:
Yeah, there is complete overlap. Each event payload contains the same fields, but in a different format. This is the only way I could think of parsing them into a single type (`PullRequestMetadata`).